Added new popwindow module in sugar3.graphics

AbrahmAB opened the pull request:
@samdroid-apps , as discussed I have designed the pop-up model here.
Use of this can be seen in my PR in sugar repo.
Link-> https://github.com/sugarlabs/sugar/pull/672

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,146 @@
+
+class PopWindow(Gtk.Window):
+    """
+    UI interface for activity Pop-up Windows
```
Please write some docs for this.  Here is an example of the format:  https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/256/files#diff-74308e345a99a21f5fbfc47f0dbcc070R106

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,146 @@
+    def _add_events(self, event):
+        pass
+
+    def add_view(self, view, expand=True, fill=True, padding=0):
```
Please document this method

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,146 @@
+                display, self._parent_window_xid)
+            window.set_transient_for(parent)
+
+    def _set_decorated(self, decorated):
```
Can't you just use the get_decorated and set_decorated methods that the Gtk.Window already has?  Same for the position, border_width, resize_grip.

AbrahmAB replied:
This method can be further used by other class which inherits this class to set_decorated according to their needs.
Same is for position, border_width, resize_grip

samdroid-apps replied:
Can't the sub classes just use the normal "set_decorated" menthods?  Why do we need to keep a copy of the value as "self._decorated"?  Can't you just set with "set_decorated" and get with "get_decorated"?

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,146 @@
+        self._has_resize_grip = has_resize_grip
+        self.set_has_resize_grip(self._has_resize_grip)
+
+    def _add_events(self, event):
```
What is this for?

AbrahmAB replied:
To add gtk events.

quozl commented:
Please add more explanation to the commit description and force push.  I've no idea what the intention of the patch is.

quozl commented on src/sugar3/graphics/Makefile.am:
```diff
@@ -16,6 +16,7 @@ sugar_PYTHON =              \
 	palettemenu.py          \
 	palettewindow.py        \
 	panel.py                \
+        popwindow.py            \
```
Probably mismatch of tab and space characters in this line compared to the lines either side, causing diff to show slight indentation.

AbrahmAB commented:
@samdroid-apps wrote docs for class and methods implemented.
@quozl commit message changed.

quozl commented:
@AbrahmAB, thanks for changing the commit message, but perhaps I didn't say enough about it.

You wrote:
> Refactoring the Pop-up windows in sugar, implemented a basic modal for pop-up windows

* the patch looks like it adds a new module, but the message says it is a refactoring, (as if you mixed up the commit message with https://github.com/sugarlabs/sugar/pull/672),
* there's no link in the commit message to the corresponding commit hash in the pull request of the other repository,
* the first line of the message is too long, it wraps in displays, please use a shorter first line, then a blank line, then a paragraph of narrative or bullet points if necessary to explain,
* *sugar* is redundant information, we know from context that this is a patch for Sugar,
* look for ways to shorten the text, e.g. *refactoring* could be *refactor*, and *pop-up windows* need only be mentioned once.

Hope that helps!  The purpose of commit messages is to communicate to other programmers now and in the future.  Programmers who are doing a bisection to find the cause of a regression need as much information about the change as possible, yet in a form that is readily taken up.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,269 @@
+    These pop-up windows don't cover the whole screen.
+    They contain canvas content, alerts messages, a tray and a
+    toolbar.
+    """
```
Please document the constructor kwargs in this section, eg:

```
Kwargs:
   title (type): what is it 
   set_titlebox (type): what is it 
   window_xid (type): what is it

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,269 @@
+
+    def get_vbox(self):
+        '''
+        Getter method for self._vbox
```
The idea of docs is that I only have to read the docs and do not need to read the code.

`self._vbox` means nothing to me.  You need to tell me where the vbox is; is it the toolbar?  is it the canvas?

Please do this for the get_scroll_window function as well.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,269 @@
+        '''
+        self._events = event
+        if self._events is not None:
+            self.add_events(self._events)
```
Cant we just use add_events and get_events here?

Also, the way that you are maintaining the `_events` variable is broken.  You are overwritting, no appending (this is add events, not set events).  Please just use add_events and get_events.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,269 @@
+        '''
+        return self._sample_window
+
+    def __close_button_clicked_cb(self, button):
```
callbacks with two leading underscores are very private.  They don't really need docs.

AbrahmAB replied:
So should I remove the docs for all very private callbacks?

AbrahmAB commented:
@samdroid-apps 
* Added document the constructor kwargs for PopWindow class.
* Improved docs for get_vbox and get_scroll_window.
* Replaced variable `_sample_window` with `_scroll_window`.
* Removed the _add_events and copy of Gtk.Window methods like self._set_decorated etc.
* Changed the method name `_add_widget` to `add_widget` in TitleBox.

AbrahmAB commented:
@quozl Thanks for your pointing out main points while writing a commit message.
I have changed the commit message again. Hope, that would be more informative! :)

AbrahmAB commented:
Oops! I made some typo error while changing the commit message of my PR in sugar repo having reference of this PR, for which I had to `commit --amend -m` again. I apologize for  multiple reference links above :(.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
```
Can we have an example of using this at the top of the file.  This would be more like the documentation standard that we want to achive.

Maybe look at the example from an activity perspective?  For example, I plan to use this API in Bibliography to fix up the modal "new item" winodow.  Read activity will also use it for the bookmark window.

AbrahmAB replied:
Done! :)

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+    toolbar.
+
+    Kwargs:
+        title (gettext.gettext or _() ): title label of the pop-up window
```
don't you want a string here

AbrahmAB replied:
As per the recent changes title of the popwindow is made GObject property, there is no need of passing title argument for now while initializing the PopWindow class.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+            self._title_box.show()
+
+        width = Gdk.Screen.width() - style.GRID_CELL_SIZE * 5
+        height = Gdk.Screen.height() - style.GRID_CELL_SIZE * 3
```
We really should provide an api to add sizes.  You could use add a GObject property to control them.

Maybe also include some constants for the sizes, like:

`PopWindowSize.FULLSCREEN` = like the view source window
`PopWindowSize.HALF_WIDTH` = like the bibliography "add item" window

AbrahmAB replied:
Done!
provided to variable `FULLSCREEN ` and `HALF_WIDTH` for setting size as required.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+    """
+
+    def __init__(self, title=None, set_titlebox=True, window_xid=None):
+        Gtk.Window.__init__(self)
```
Please proxy the kword argments to the constructor:

```
def __init__(self, title=None, set_titlebox=True, window_xid=None, **kwargs):
    Gtk.Window.__init__(self, **kwargs)
```

Since the GtkWindow is a GOBject, that lets me go:

```
p = PopWindow(transistent_for=self)
```

rather than

```
p = PopWindow()
p.props.transistent_for = self

AbrahmAB replied:
Done! :)

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+        self.close_button.show()
+
+        label = Gtk.Label()
+        label.set_markup('<b>%s</b>' % title)
```
Can we make the title into a GObject property?  That would let me change the title of the titlebox - I use this feature in biblography activity "browse import" window.

You can use the GObject.Property api to do this:  http://python-gtk-3-tutorial.readthedocs.org/en/latest/objects.html#properties

AbrahmAB replied:
Yes done :)
Have a look at the example of this https://github.com/sugarlabs/sugar/pull/672/files#diff-52320dfca28b62095ceb6c9c63c87b9dR322

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+        self.connect('key-press-event', self.__key_press_event_cb)
+        self.connect('hide', self.__hide_cb)
+
+        self._vbox = Gtk.VBox()
```
Please use Gtk.Box(orientation=Gtk.Orientation.VERTICAL)

Gtk.VBox is deprecated :(

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,204 @@
+        self._vbox.show()
+
+        if self._set_title_box:
+            self._title_box = TitleBox(title)
```
I went to start porting ViewHelp to use this.  But there is no public api to get the titlebox :(

Please add like a get_titlebox method or add a titlebox property

AbrahmAB replied:
Added `get_title_box` method

quozl commented:
Reviewed.  Agree with comments by @samdroid-apps.

AbrahmAB commented:
@samdroid-apps Changes have been made as per your suggestions.

quozl commented:
Reviewed.

samdroid-apps commented:
The api looks good.  I'm afk so I cant test it, but I'm sure that @quozl can merge it if he is satisfied.

quozl commented:
Happy to merge into my fork, but I don't think I have community support to merge into Sugar Labs.

i5o commented:
I will test ASAP.


Hi everyone btw.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,245 @@
+    HALF_WIDTH = ((Gdk.Screen.height() - style.GRID_CELL_SIZE * 5)/2,
+                  (Gdk.Screen.height() - style.GRID_CELL_SIZE * 3))
+
+    def __init__(self, set_titlebox=True, window_xid=None, **kwargs):
```
This patch is looking great.  Just 2 little changes that might polish it up:

1.  Could we remove the "set_titlebox" property?  Most people will just use the titlebox, and it can always be removed via `get_titlebox().hide()`.

2.  Could you add a property or constructor argument that takes one of the size constants?

AbrahmAB replied:
Adding props for size(having tuple of width, height)  or separate for width and height.
I would prefer the separate properties. What are your suggestions @samdroid-apps ?

samdroid-apps replied:
Having just one prop would be very good developer experience:

```
popup = PopWindow(PopWindow.FULLSCREEN)
```

On Thu, Apr 21, 2016 at 7:55 AM, Abhijit Patel 
<notifications@github.com> wrote:
> In src/sugar3/graphics/popwindow.py:
> 
> > +    They contain canvas content, alerts messages, a tray and a
> > +    toolbar.
> > +
> > +    FULLSCREEN and HALF_WIDTH for setting size of the window.
> > +
> > +    Kwargs:
> > +        set_titlebox (bool): False if titlebox is not needed
> > +        window_xid (xlib.Window): xid of the parent window
> > +    """
> > +    FULLSCREEN = (Gdk.Screen.width() - style.GRID_CELL_SIZE * 5,
> > +                  Gdk.Screen.height() - style.GRID_CELL_SIZE * 3)
> > +
> > +    HALF_WIDTH = ((Gdk.Screen.height() - style.GRID_CELL_SIZE * 
> 5)/2,
> > +                  (Gdk.Screen.height() - style.GRID_CELL_SIZE * 3))
> > +
> > +    def __init__(self, set_titlebox=True, window_xid=None, 
> **kwargs):
> Adding props for size(having tuple of width, height) or separate for 
> width and height.
> I would prefer the separate properties. What are your suggestions 
> @samdroid-apps ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

AbrahmAB commented:
Changes made:
1. Removed set_title_box property.
2. Added constructor argument `size` .

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,241 @@
+    def __init__(self):
+        ToolbarBox.__init__(self)
+
+        self.close_button = ToolButton(icon_name='dialog-cancel')
```
I was just testing as I was integrating this into bibliography activity.  Can you set the "tooltip=_('Close')", to be consistent with existing popups?

AbrahmAB replied:
`tooltip` set :+1:

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,241 @@
+    '''
+    Title box at the top of the pop-up window.
+    Title and close button are added to the box and as needed more widgets
+    can be added using self.add_widget method.
```
Is there a way to add the a widget to the left of the close button?  Eg, I would like a toolbar like:

    Edit Bibliography Entry [Add] [Close]

But when I use `self.get_title_box().toolbar.insert(add, -2)` or `self.get_title_box().toolbar.insert(add, -1)`, I get:

[![https://gyazo.com/a053b8d5b73b8624cad34507dba7f049](https://i.gyazo.com/a053b8d5b73b8624cad34507dba7f049.png)](https://gyazo.com/a053b8d5b73b8624cad34507dba7f049)

(Which has the add and close buttons in the wrong order)

AbrahmAB replied:
Fixed the order of buttons @samdroid-apps :)
changed `self.toolbar.insert(self.close_button, 0)` to `self.toolbar.insert(self.close_button, -1)`
Now for 
``` 
Edit Bibliography Entry [Add] [Close]
```
you can use self.get_title_box().toolbar.insert(add, 1)

samdroid-apps commented:
I will do a code review soon, but 1st there are some issues to fix:

![screenshot from 2016-05-24 16-28-34](https://cloud.githubusercontent.com/assets/6022042/15494235/d7969728-2178-11e6-96e6-c2007d2473b5.png)
Toolbar adding padding - the label should be on the very right and the buttons on the very left

![screenshot from 2016-05-24 16-25-26](https://cloud.githubusercontent.com/assets/6022042/15494237/d82e441a-2178-11e6-9b68-d07c8b12c20e.png)
Left buttons are in the wrong order, padding is an issue

![screenshot from 2016-05-24 16-24-12](https://cloud.githubusercontent.com/assets/6022042/15494236/d7de3be6-2178-11e6-8f04-06903a9b008b.png)
Left buttons are in the wrong order, padding is an issue

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+        # the child class to display the window after modifications
+        # like chaninging window size, decorating, changing position.
+
+        #self.show()
```
Please remove this commented code

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+
+        Returns:
+            None
+        '''
```
Please remove the docstrings on these private callback functions (`__*`).

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+        self._label.set_markup('<b>%s</b>' % title)
+        self._label.show()
+
+    title = GObject.property(type=str, setter=set_title)
```
`PyGIDeprecationWarning: GObject.property is deprecated; use GObject.Property instead` :(

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+
+    def add_widget(self, widget, expand=False, position=-1):
+        '''
+        Adds widget to the title-box.
```
Document that this func wraps it in a toolitem, and that you can just insert buttons via `x.props.toolbar.insert(...)`

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+
+        #self.show()
+
+    def get_title_box(self):
```
Can this be a gobject prop?

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+    HALF_WIDTH = ((Gdk.Screen.height() - style.GRID_CELL_SIZE * 5)/2,
+                  (Gdk.Screen.height() - style.GRID_CELL_SIZE * 3))
+
+    def __init__(self, size=FULLSCREEN, window_xid=None, **kwargs):
```
Can we make the size into a gobject property?  `pydoc gi.repository.GObject.Property`

Once you made it into a prop, then it will be automatically set via the **kwargs thing.

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+        '''
+        return self._title_box
+
+    def get_vbox(self):
```
Can this be a gobject prop?

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+#   Physics activity for Load Sample window.
+#   Read activity for the bookmark window.
+#   Bibliography for "new item" winodow.
+#   Pippy activity for Load example window.
```
No need for a list of examples in this code comment

samdroid-apps commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,244 @@
+#   Read activity for the bookmark window.
+#   Bibliography for "new item" winodow.
+#   Pippy activity for Load example window.
+
```
Can you add a docstring with an example in here?

Maybe just add a simple example:  making a PopWindow containing a GtkLabel and with an extra toolbutton in the toolbar.

samdroid-apps replied:
For how to do this doc syntax wise, see src/sugar3/graphics/radiotoolbutton.py  line 19

AbrahmAB commented:
Forced push with changes:
* Fixed padding issue :) used `Gtk.Toolbar` instead of `sugar3.graphics.ToolbarBox`
* Removed commented codes
* Added docs string with example as per the syntax
* `GObject.Property` for `size` , `title_box`, `vbox`, `TitleBox.title`
* Removed `add_widget` method in `TitleBox` class

samdroid-apps commented:
Awsome!  Merged!  I can't wait to refactor my activities to use this.

iamutkarshtiwari commented on src/sugar3/graphics/popwindow.py:
```diff
@@ -0,0 +1,202 @@
+        # like chaninging window size, decorating, changing position.
+
+    def set_size(self, size):
+        width, height = size
```
This way of assigning two variable with one value isn't right. It should be somewhat like this ```width, height = size_x, size_y```

samdroid-apps replied:
No, you pass a tuple;

    tuple = (1, 200)
    a, b = tuple

AbrahmAB replied:
The argument for this function is tuple of size itself!

iamutkarshtiwari replied:
Ah.. Got it

iamutkarshtiwari commented:
I am using this module for my 'Screenshot' popup. But whenever I click a snapshot, before the popup shows up on the screens, I see a ```grey``` rectange appear for a small fraction of second (which also get's captured in my screenshot'). You can see it in the attached screenshot below -
![desktop-animation](https://cloud.githubusercontent.com/assets/6258810/16073575/380df942-3304-11e6-8500-88e1ea2a4c49.gif)

Captured screenshot below - 

![data](https://cloud.githubusercontent.com/assets/6258810/16073605/6463b72a-3304-11e6-99ab-be0013b00446.png)

quozl commented:
Possibly a window realisation and size negotiation race.

The early stages of `PopWindow.__init__` may be forcing Gtk to create the Gdk window in order to set the attributes requested.  However, at the time this happens, there are no children in the window to negotiate a size.  The children are added later.

You might experiment by moving parts of `PopWindow.__init__` around; add the child widgets as soon as possible.  Then call `self.connect` for the `realize` event.  Then make the other calls.  Then call `self.connect` for the remaining events.

Or, perhaps @AbrahmAB's design feature of not calling `self.show` (see comments at end of `__init__`) may need to be revisited.

AbrahmAB commented:
Hi Utkarsh,
I tried testing your patch but its showing the following traceback in my log
`Traceback (most recent call last):`
`  File "/home/broot/sugar-build/build/out/install/lib/python2.7/site-packages/jarabe/frame/devicestray.py", line 34, in __init__`
    `locals(), [module_name])`
`  File "/home/broot/sugar-build/build/out/install/share/sugar/extensions/deviceicon/display.py", line 37, in <module>`
   ` from jarabe.screenshotpanel.gui import ScreenshotPanel`
`ImportError: No module named screenshotpanel.gui`

quozl commented:
@AbrahmAB, the error shows file `gui.py` is missing; check that you have properly applied the patches in the pull request and that a file `gui.py` is in the `screenshotpanel` directory.  See how https://github.com/sugarlabs/sugar/pull/675/commits/02636deaebc406b04db4fcefd723ea3781fb3778 adds that file.
